### PR TITLE
Complete deduction tests, implement guides for all containers

### PR DESCRIPTION
**Pull request by cmazakas**
Closes https://github.com/boostorg/unordered/issues/149

### CODE REVIEW

File: include/boost/unordered/unordered_flat_map.hpp [hunk 1]
```diff
@@ -628,6 +628,104 @@ namespace boost {
...
+    namespace detail {
+      template <typename T>
+      using iter_key_t =
+        typename std::iterator_traits<T>::value_type::first_type;
+      template <typename T>
+      using iter_val_t =
+        typename std::iterator_traits<T>::value_type::second_type;
+      template <typename T>
+      using iter_to_alloc_t =
+        typename std::pair<iter_key_t<T> const, iter_val_t<T> >;
```
Comment (pdimov) [inline include/boost/unordered/unordered_flat_map.hpp:642]: I don't think `typename` is needed here.

**Comment by joaquintides** [comment]
Why's the `boost::type_identity` trick only used for maps and not for sets?

**Comment by pdimov** [comment]
This passed CI; should I merge it to develop and master? We can fix the remaining things later.

**Comment by joaquintides** [comment]
I'm OK with that. Docs should be updated --I can do that, but, again, not this week

**Pull request merged**